File: transfomer/canvas_border_autocrop/grower.py
"""Seeded 4-connected background region growing (OpenCV-accelerated)."""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass

import cv2
import numpy as np

from .config import DetectorConfig
from .similarity import BackgroundSimilarity
from .types import IntRect, SeedPatch, SemanticSide, SIDE_INDEX


@dataclass(slots=True)
class GrownBackground:
    mask: np.ndarray  # bool HxW
    source_label: np.ndarray  # int8 HxW, 0=none, 1..4 = sides
    pixel_count: int
    bbox_fill_ratio: float = 1.0
    hole_score: float = 0.0
    touches_search_border: float = 0.0


def grow_background(
    seeds: list[SeedPatch],
    model_seed_ids: set[int],
    similarity: BackgroundSimilarity,
    search_roi: IntRect,
    cfg: DetectorConfig,
) -> GrownBackground | None:
    h, w = similarity.similarity.shape
    x0, x1 = search_roi.left, search_roi.right
    y0, y1 = search_roi.top, search_roi.bottom
    if x1 <= x0 + 2 or y1 <= y0 + 2:
        return None

    start_seeds = [
        s
        for s in seeds
        if s.accepted and s.seed_id in model_seed_ids and search_roi.contains_point(s.x, s.y)
    ]
    if not start_seeds:
        start_seeds = [
            s for s in seeds if s.seed_id in model_seed_ids and search_roi.contains_point(s.x, s.y)
        ]
    if not start_seeds:
        return None

    strong = similarity.strong_mask
    weak = similarity.weak_mask
    barrier = similarity.barrier_mask

    # Growable: weak similarity, stopped by sustained barriers unless strong
    growable = weak.copy()
    growable[barrier & ~strong] = False
    growable[:y0, :] = False
    growable[y1:, :] = False
    growable[:, :x0] = False
    growable[:, x1:] = False

    # Bridge ≤ grow_gap_bridge_px isolated noise without filling large canvas holes
    gap = max(0, min(cfg.grow_gap_bridge_px, 2))
    sub = growable[y0:y1, x0:x1]
    if gap > 0 and sub.size:
        k = 2 * gap + 1
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (k, k))
        closed = cv2.morphologyEx(sub.astype(np.uint8), cv2.MORPH_CLOSE, kernel)
        # Only keep closed pixels that stay within weak band after bridging
        weak_sub = weak[y0:y1, x0:x1]
        strong_sub = strong[y0:y1, x0:x1]
        bridged = (closed > 0) & (weak_sub | strong_sub)
        growable[y0:y1, x0:x1] = bridged

    # Drop seeds that are not growable
    valid_seeds = [s for s in start_seeds if growable[s.y, s.x]]
    if not valid_seeds:
        return None

    nlab, labels = cv2.connectedComponents(growable.astype(np.uint8), connectivity=4)
    if nlab <= 1:
        return None

    keep_labels: set[int] = set()
    label_sides: dict[int, list[int]] = {}
    for s in valid_seeds:
        lid = int(labels[s.y, s.x])
        if lid <= 0:
            continue
        keep_labels.add(lid)
        label_sides.setdefault(lid, []).append(SIDE_INDEX[s.side] + 1)

    if not keep_labels:
        return None

    mask = np.zeros((h, w), dtype=bool)
    source = np.zeros((h, w), dtype=np.int8)
    count = 0
    max_pixels = int((x1 - x0) * (y1 - y0) * cfg.max_grown_fraction)

    for lid in keep_labels:
        comp = labels == lid
        c = int(np.count_nonzero(comp))
        if count + c > max_pixels:
            # Keep largest seed-touched components first
            continue
        mask |= comp
        maj = Counter(label_sides.get(lid, [1])).most_common(1)[0][0]
        source[comp] = maj
        count += c

    if count < 16:
        return None

    ys, xs = np.nonzero(mask)
    bx0, bx1 = int(xs.min()), int(xs.max()) + 1
    by0, by1 = int(ys.min()), int(ys.max()) + 1
    bbox_area = max(1, (bx1 - bx0) * (by1 - by0))
    fill_ratio = float(count) / float(bbox_area)
    hole_score = float(np.clip(1.0 - fill_ratio, 0.0, 1.0))

    # Fraction of search-roi border pixels that are grown (UI chrome leaks)
    border_hits = 0
    border_tot = 0
    for y in (y0, y1 - 1):
        if 0 <= y < h:
            row = mask[y, x0:x1]
            border_hits += int(np.count_nonzero(row))
            border_tot += row.size
    for x in (x0, x1 - 1):
        if 0 <= x < w:
            col = mask[y0:y1, x]
            border_hits += int(np.count_nonzero(col))
            border_tot += col.size
    touch = border_hits / max(border_tot, 1)

    return GrownBackground(
        mask=mask,
        source_label=source,
        pixel_count=count,
        bbox_fill_ratio=fill_ratio,
        hole_score=hole_score,
        touches_search_border=float(touch),
    )


def model_geometry_plausible(grown: GrownBackground, cfg: DetectorConfig) -> bool:
    """Hard filter: workspace bg is typically a frame/L/bands with a canvas hole.

    Solid fills that hug the search border are usually UI chrome.
    Solid fills that do not hug the border are usually the canvas itself.
    """
    # Large hole / frame → good
    if grown.hole_score >= cfg.min_bg_hole_score:
        return True
    # Thin bands may have moderate fill if bbox is tight around strips
    if grown.bbox_fill_ratio <= cfg.max_bg_bbox_fill_ratio and grown.touches_search_border < 0.55:
        return True
    # Solid + heavy search-border contact → UI
    if grown.touches_search_border >= cfg.max_bg_search_border_touch:
        return False
    # Solid interior blob → likely canvas
    if grown.bbox_fill_ratio >= 0.85 and grown.hole_score < 0.12:
        return False
    return grown.hole_score >= cfg.min_bg_hole_score * 0.5

File: screen_canvas_transform/native/src/grower.cpp
#include "wb/grower.hpp"

#include <algorithm>
#include <queue>
#include <unordered_set>

namespace wb {
namespace {

bool ContainsId(const std::vector<int>& ids, int id) {
  return std::find(ids.begin(), ids.end(), id) != ids.end();
}

void BridgeGaps(ImageU8& growable, const ImageU8& weak, const ImageU8& strong, const IntRect& roi,
                int gap) {
  if (gap <= 0) return;
  const int x0 = roi.left, x1 = roi.right, y0 = roi.top, y1 = roi.bottom;
  ImageU8 dilated;
  dilated.Allocate(growable.width, growable.height, 0);
  for (int y = y0; y < y1; ++y) {
    for (int x = x0; x < x1; ++x) {
      if (!growable.At(x, y)) continue;
      for (int dy = -gap; dy <= gap; ++dy) {
        for (int dx = -gap; dx <= gap; ++dx) {
          const int xx = x + dx;
          const int yy = y + dy;
          if (xx < x0 || xx >= x1 || yy < y0 || yy >= y1) continue;
          dilated.At(xx, yy) = 1;
        }
      }
    }
  }
  for (int y = y0; y < y1; ++y) {
    for (int x = x0; x < x1; ++x) {
      const bool keep = dilated.At(x, y) && (weak.At(x, y) || strong.At(x, y));
      growable.At(x, y) = keep ? 1 : 0;
    }
  }
}

}  // namespace

GrownBackground* GrowBackground(const std::vector<SeedPatch>& seeds,
                                const std::vector<int>& model_seed_ids,
                                const BackgroundSimilarity& similarity, const IntRect& grow_roi,
                                const DetectorConfig& cfg, GrownBackground& out) {
  const int h = similarity.similarity.height;
  const int w = similarity.similarity.width;
  const int x0 = grow_roi.left;
  const int x1 = grow_roi.right;
  const int y0 = grow_roi.top;
  const int y1 = grow_roi.bottom;
  if (x1 <= x0 + 2 || y1 <= y0 + 2) return nullptr;

  std::vector<const SeedPatch*> start_seeds;
  for (const auto& s : seeds) {
    if (s.accepted && ContainsId(model_seed_ids, s.seed_id) && grow_roi.ContainsPoint(s.x, s.y))
      start_seeds.push_back(&s);
  }
  if (start_seeds.empty()) {
    for (const auto& s : seeds) {
      if (ContainsId(model_seed_ids, s.seed_id) && grow_roi.ContainsPoint(s.x, s.y))
        start_seeds.push_back(&s);
    }
  }
  if (start_seeds.empty()) return nullptr;

  ImageU8 growable;
  growable.Allocate(w, h, 0);
  for (int y = y0; y < y1; ++y) {
    for (int x = x0; x < x1; ++x) {
      // Keep weak/strong background pixels growable even on gradient ridges so the
      // outer rim (bg↔UI transition) is not shaved off. Barriers still stop leakage
      // into non-background neighbors during flood fill below.
      const bool g = similarity.weak_mask.At(x, y) != 0 || similarity.strong_mask.At(x, y) != 0;
      growable.At(x, y) = g ? 1 : 0;
    }
  }
  BridgeGaps(growable, similarity.weak_mask, similarity.strong_mask, grow_roi,
             std::max(0, std::min(cfg.grow_gap_bridge_px, 2)));

  std::vector<const SeedPatch*> valid;
  for (auto* s : start_seeds) {
    if (growable.At(s->x, s->y)) valid.push_back(s);
  }
  if (valid.empty()) return nullptr;

  out.mask.Allocate(w, h, 0);
  out.source_label.Allocate(w, h, 0);
  ImageU8 visited;
  visited.Allocate(w, h, 0);

  // No hard area cap vs full capture: stop via barriers / non-bg. Soft safety only.
  const int roi_area = std::max(1, (x1 - x0) * (y1 - y0));
  const int max_pixels = (cfg.max_grown_fraction > 0.f)
                             ? static_cast<int>(roi_area * cfg.max_grown_fraction)
                             : roi_area;
  int count = 0;

  // Grow components from each seed; merge overlapping
  struct Comp {
    int pixels = 0;
    int side_vote[5] = {};
  };
  std::vector<Comp> comps;
  ImageU8 labels;
  labels.Allocate(w, h, 0);
  int next_label = 1;

  auto flood = [&](int sx, int sy, int side_idx) {
    if (!growable.At(sx, sy) || labels.At(sx, sy)) return;
    Comp c;
    std::queue<std::pair<int, int>> q;
    q.push({sx, sy});
    labels.At(sx, sy) = static_cast<uint8_t>(next_label);
    c.side_vote[side_idx]++;
    c.pixels = 1;
    static const int dxs[4] = {1, -1, 0, 0};
    static const int dys[4] = {0, 0, 1, -1};
    while (!q.empty()) {
      const auto p = q.front();
      q.pop();
      for (int k = 0; k < 4; ++k) {
        const int nx = p.first + dxs[k];
        const int ny = p.second + dys[k];
        if (nx < x0 || nx >= x1 || ny < y0 || ny >= y1) continue;
        if (!growable.At(nx, ny) || labels.At(nx, ny)) continue;
        // Do not leak across strong barriers into non-strong background.
        if (similarity.barrier_mask.At(nx, ny) && !similarity.strong_mask.At(nx, ny) &&
            !similarity.weak_mask.At(nx, ny))
          continue;
        labels.At(nx, ny) = static_cast<uint8_t>(next_label);
        ++c.pixels;
        q.push({nx, ny});
      }
    }
    comps.push_back(c);
    ++next_label;
  };

  for (auto* s : valid) {
    flood(s->x, s->y, SideIndex(s->side) + 1);
  }
  if (comps.empty()) return nullptr;

  // Keep seed-touched labels in order of size
  std::vector<int> order(comps.size());
  for (size_t i = 0; i < comps.size(); ++i) order[i] = static_cast<int>(i);
  std::sort(order.begin(), order.end(),
            [&](int a, int b) { return comps[a].pixels > comps[b].pixels; });

  std::unordered_set<int> keep;
  for (int i : order) {
    if (count + comps[i].pixels > max_pixels && count > 0) continue;
    keep.insert(i + 1);
    count += comps[i].pixels;
  }
  if (count < 16) return nullptr;

  int bx0 = w, bx1 = 0, by0 = h, by1 = 0;
  for (int y = y0; y < y1; ++y) {
    for (int x = x0; x < x1; ++x) {
      const int lid = labels.At(x, y);
      if (!lid || !keep.count(lid)) continue;
      out.mask.At(x, y) = 1;
      int maj = 1, majv = -1;
      for (int s = 1; s <= 4; ++s) {
        if (comps[lid - 1].side_vote[s] > majv) {
          majv = comps[lid - 1].side_vote[s];
          maj = s;
        }
      }
      out.source_label.At(x, y) = static_cast<uint8_t>(maj);
      bx0 = std::min(bx0, x);
      bx1 = std::max(bx1, x + 1);
      by0 = std::min(by0, y);
      by1 = std::max(by1, y + 1);
    }
  }

  out.pixel_count = count;
  out.bbox = {bx0, by0, bx1, by1};
  const int bbox_area = std::max(1, (bx1 - bx0) * (by1 - by0));
  out.bbox_fill_ratio = static_cast<float>(count) / static_cast<float>(bbox_area);
  out.hole_score = std::max(0.f, std::min(1.f, 1.f - out.bbox_fill_ratio));

  int border_hits = 0, border_tot = 0;
  for (int y : {y0, y1 - 1}) {
    if (y < 0 || y >= h) continue;
    for (int x = x0; x < x1; ++x) {
      ++border_tot;
      if (out.mask.At(x, y)) ++border_hits;
    }
  }
  for (int x : {x0, x1 - 1}) {
    if (x < 0 || x >= w) continue;
    for (int y = y0; y < y1; ++y) {
      ++border_tot;
      if (out.mask.At(x, y)) ++border_hits;
    }
  }
  out.touches_capture_border = border_hits / static_cast<float>(std::max(border_tot, 1));
  return &out;
}

bool IsWorkspaceBackgroundModel(const GrownBackground& grown, const BackgroundModel& model,
                                const DetectorConfig& cfg) {
  // Solid canvas-like blob: high fill, almost no interior hole.
  if (grown.bbox_fill_ratio >= 0.85f && grown.hole_score < 0.12f) return false;
  if (grown.bbox_fill_ratio > cfg.max_bg_bbox_fill_ratio &&
      grown.hole_score < cfg.min_bg_hole_score) {
    return false;
  }
  // Weak multi-side rectangular support → likely interior canvas color.
  if (model.rectangular_support_score < cfg.min_model_rect_support) return false;

  if (grown.hole_score >= cfg.min_bg_hole_score) return true;
  if (grown.bbox_fill_ratio <= cfg.max_bg_bbox_fill_ratio &&
      grown.touches_capture_border < 0.55f) {
    return true;
  }
  // Touching capture border heavily without a hole looks like leaked UI / solid panel.
  if (grown.touches_capture_border >= cfg.max_bg_capture_border_touch) return false;
  return grown.hole_score >= cfg.min_bg_hole_score * 0.5f;
}

}  // namespace wb
